cubed/icechunk.py

```diff
@@ -0,0 +1,86 @@
+from typing import TYPE_CHECKING, Any, List, Sequence, Union
+
+import zarr
+from icechunk import IcechunkStore
+
+from cubed import compute
+from cubed.core.array import CoreArray
+from cubed.core.ops import blockwise
+from cubed.runtime.types import Callback
+
+if TYPE_CHECKING:
+    from cubed.array_api.array_object import Array
+
+
+def store_icechunk(
+    store: IcechunkStore,
+    *,
+    sources: Union["Array", Sequence["Array"]],
+    targets: List[zarr.Array],
+    executor=None,
+    **kwargs: Any,
+) -> None:
+    if isinstance(sources, CoreArray):
+        sources = [sources]
+        targets = [targets]
+
+    if any(not isinstance(s, CoreArray) for s in sources):
+        raise ValueError("All sources must be cubed array objects")
+
+    if len(sources) != len(targets):
+        raise ValueError(
+            f"Different number of sources ({len(sources)}) and targets ({len(targets)})"
+        )
+
+    if isinstance(sources, CoreArray):
+        sources = [sources]
+        targets = [targets]
+
+    arrays = []
+    for source, target in zip(sources, targets):
+        identity = lambda a: a
+        ind = tuple(range(source.ndim))
+        array = blockwise(
+            identity,
+            ind,
+            source,
+            ind,
+            dtype=source.dtype,
+            align_arrays=False,
+            target_store=target,
+            return_writes_stores=True,
+        )
+        arrays.append(array)
+
+    # use a callback to merge icechunk stores
+    store_callback = IcechunkStoreCallback()
+    # add to other callbacks the user may have set
+    callbacks = kwargs.pop("callbacks", [])
+    callbacks = [store_callback] + list(callbacks)
+
+    compute(
+        *arrays,
+        executor=executor,
+        _return_in_memory_array=False,
+        callbacks=callbacks,
+        **kwargs,
+    )
+
+    # merge back into the store passed into this function
+    merged_store = store_callback.store
+    store.merge(merged_store.change_set_bytes())
+
+
+class IcechunkStoreCallback(Callback):
+    def on_compute_start(self, event):
+        self.store = None
+
+    def on_task_end(self, event):
+        result = event.result
+        if result is None:
+            return
+        for store in result:
+            if self.store is None:
+                self.store = store
+            else:
+                self.store.merge(store.change_set_bytes())
```

Comment:
@paraseba shall we commit to `icechunk.distributed.merge_stores` as public API?

https://github.com/earth-mover/icechunk/blob/a8c33b81d329c32a8e34e7151a1a71620967067a/icechunk-python/python/icechunk/distributed.py#L12-L16

Reply:
@tomwhite It looks like callbacks are "accumulated" on every worker. Is that right?

Reply:
Right we can fix that, but note that dask and presumably cubed are accumulating on remote workers already,  so there is already some parallelism in how it is used.

Moreover it'd be nice not to have `store.change_set_bytes` be the public API :)

Reply:
> but would it be possible to merge a batch in one go rather than one at a time

we'll have to build some rust API, we'll get to this eventually.

> No, they are being accumulated on the client so there is no parallelism. 

ah ok. In that case, how about calling `reduction` on each array that was written? That way you parallelize the merge across blocks for each array, and then the only serial bit is the merging across arrays, which will be a lot smaller. I considered this approach for dask, but then just wrote out a tree reduction across all chunks.

Reply:
> Yes, that's basically it

:( I was afraid so.